File: template-method/kotlin/DataAnalyzerBase.kt
package TemplateMethod

// Metot iskeletimizi ve soyutlanmış metotları içermektedir.
// UML diyagramındaki AbstractClass yapısına denk gelmektedir.
internal abstract class DataAnalyzerBase {

    // Burada veri verinin çağrılma süreçleri yer alacaktır.
    // Default olarak bir obje dönülmektedir.
    private fun getDataOfLastWeek(): Any {
        println("Method worked: GetDataOfLastWeek")
        return Any()
    }

    // Metot iskeletimiz.
    fun analyzeDataOfLastWeek() {
        val data = getDataOfLastWeek()
        val result = analyzeData(data)
        exportResult(result)
    }

    private fun analyzeData(data: Any): Any {
        // Burada verinin analizi süreci yer almaktadır.
        // Analiz sonuçları default olarak bir obje dönmektedir.
        println("Method worked: AnalyzeData")
        return Any()
    }

    // Alt sınıflar tarafından uygulanmasını istediğimiz metottur.
    // Metot iskeleti içerisinde farklılık gösterecek olan metotumuzdur.
    abstract fun exportResult(result: Any)
}

File: template-method/typescript/data-analyzer-base.ts
// Metot iskeletimizi ve soyutlanmış metotları içermektedir.
// UML diyagramındaki AbstractClass yapısına denk gelmektedir.
export abstract class DataAnalyzerBase {
  // Metot iskeletimiz.
  public analyzeDataOfLastWeek(): void {
    const data = this.getDataOfLastWeek();
    const result = this.analyzeData(data);
    this.exportResult(result);
  }

  private getDataOfLastWeek(): object {
    // Burada veri verinin çağrılma süreçleri yer alacaktır.
    // Default olarak bir obje dönülmektedir.
    console.log("Method worked: GetDataOfLastWeek");
    return {};
  }

  private analyzeData(data: object): object {
    // Burada verinin analizi süreci yer almaktadır.
    // Analiz sonuçları default olarak bir obje dönmektedir.
    console.log("Method worked: AnalyzeData");
    return {};
  }

  // Alt sınıflar tarafından uygulanmasını istediğimiz metottur.
  // Metot iskeleti içerisinde farklılık gösterecek olan metotumuzdur.
  public abstract exportResult(result: object): void
}
